R/model-wrappers.R

```diff
@@ -290,3 +290,173 @@ forecastHybrid_model <- function(y = NULL, samples = NULL,
+#'                                   stackr_mixture_model(models = models,
+#'                                                        weighting_period = 5,
+#'                                                        ...)},
+#'                                 horizon = 7, samples = 10,
+#'                                 serial_interval = example_serial_interval,
+#'                                min_points = 10)
+#'
+#' plot_forecast_evaluation(forecast_eval$forecast_rts,
+#'                          EpiSoon::example_obs_rts,
+#'                          horizon_to_plot = 7)
+#'
+#' }
+#'
+
+stackr_mixture_model <- function(y = NULL,
+                                 models = NULL,
+                                 samples = NULL,
+                                 horizon = NULL,
+                                 weighting_period = 5) {
+
+
+  #### Error Handling
+  # check if y is there
+  if(is.null(y)) stop("parameter y is missing")
+
+
+  # check if stackr is installed
+  check_suggests("stackr",
+                 dev_message = "Install using devtools::install_github('nikosbosse/stackr')")
+
+  if (length(y) <= weighting_period) {
+    stop("not enough observations to do weighting. Adjust weighting_period")
+  }
+
+  #### split data into train data and data for weighting. Use train data to
+  # generate forecasts, score them against weight data, then generate forecasts
+  # bases on the entire time series and use the obtained weights to create mixture
+  n <- length(y)
+  y_train <- y[1:(n - weighting_period)]
+  y_weight <- y[(n - weighting_period + 1):n]
+
+  #### fit models on train data and generate forecasts
+  fc_w <- purrr::map_dfr(seq_along(models),
+                         .f = function(i) {
+                           model_function <- models[[i]]
+                           out <- model_function(y = y_train,
+                                                 samples = samples,
+                                                 horizon = weighting_period)
+
+                           # bring data in the correct format the stackr package expects
+                           dplyr::as_tibble(out) %>%
+                             dplyr::mutate(sample_nr = 1:dplyr::n()) %>%
+                             tidyr::pivot_longer(names_to = "date",
+                                                 values_to = "y_pred",
+                                                 cols = -sample_nr) %>%
+                             dplyr::group_by(sample_nr) %>%
+                             dplyr::mutate(y_obs = y_weight) %>%
+                             dplyr::ungroup() %>%
+                             dplyr::mutate(model = names(models)[i],
+                                           geography = "Testland")
+
+                         })
+
+  # obtain weights based on the training forecasts generated
+  w <- stackr::stack_crps(fc_w)
+
+  #### generate real forecasts and use weights to stack
+  fc <- purrr::map_dfr(seq_along(models),
```

Comment:
this code is a copy paste of above? Suggest using an internal fn and calling twice